### ISSUE TITLE
Batch normalization

### ISSUE THREAD

**Issue by dominiksturm** [AIH-SGML/MixMIL#17]
Hey,

we've been playing around with the framework for a bit (with likelihood='binomial') and got some questions regarding the reparametrization during inference on single samples. Assuming that during inference it should be handled similar to batch normalization, we compute the mean and standard deviation from the training set and feed these as done in the `predict` function in the MixMIL module:

```python
# compute scaling for batch normalization
beta_u, beta_z = model.get_betas()
b = torch.sqrt((beta_z**2).mean(0, keepdim=True))
eta = beta_z / b

u = model._calc_bag_emb_effect_scatter(beta_u, eta, Xs)
mean, std = (u.mean(0), u.std(0))

y_pred = torch.sigmoid(model.predict(test_Xs, scaling=(mean,std)))
```

Following the discussion from #14 and given that after the GLMM initialization both the GLMM and MixMIL use the same predict function, should we give the mean and standard deviation on single samples the same way? What would be in general the best practice within the framework to perform valid comparisons to the mean GLMM? Here we observe that classification errors change quite dramatically (for the worse) with including the training set statistics, i.e. from 96% to 70% accuracy.

As an additional question how is seeding to be handled within the framework? Is there the possibility to directly supply a generator or should the generators be seeded in the general way as:
```python
np.random.seed(seed)
torch.manual_seed(seed)
```

Thanks in advance!

**Comment by jan-engelmann** [comment]
Hi Dominik,

thanks for reaching out!

Regarding seeding: yes, that's the way to go. Providing a generator would be nicer but seeding the libraries does the job.

For you question regarding the training set statistics I'd need to know a bit more about your use-case:
- What type of data are you working with? Which modality? How do you derive embeddings? What is a bag? What is an instance?
- How do you perform your train/test split? Random? Different datasets? Is there a distribution shift/batch effect between the splits?
- The Binomial likelihood is for the regression setting. If you're measuring accuracy you should consider using the categorical likelihood.

Pragmatically, I'd say if including the train statistics deteriorates your results, do not use them. The comparison to the GLMM would still be fair. A note: we have included the reparameterization because we observed it to improve training stability in some experiments. It should not affect the ranking of bags.

Hope that helps! Looking forward to hearing about your use case:)

Cheers,
Jan

**Comment by dominiksturm** [comment]
Hey Jan, 

thank you for the answer! We observe that generally performance deteriorates for the mean model when using training statistics. For the trained MixMIL model it actually improves and reduces some of the false positives. 

> What type of data are you working with? Which modality? How do you derive embeddings? What is a bag? What is an instance?

We are performing patient-level classifications, where a bag comprises different sections of a tissue. Generally, we observe very low witness rates, because of which your framework sounds very appealing. Instance embeddings are given by some sort of morphometric features directly derived from the images, for which we observe a shift in distribution between classes on the image level.

> How do you perform your train/test split? Random? Different datasets? Is there a distribution shift/batch effect between the splits?

Currently these are given by fixed datasets. In the mean time we have realized, that there is a small shift in distribution mainly due to the low data available at the moment (n=15), which we tried to augment by bag subsampling (which of course does not alleviate this effect).

> The Binomial likelihood is for the regression setting. If you're measuring accuracy you should consider using the categorical likelihood.

This I don't understand. For `n_trials=1` this should correspond to a logistic regression, right? Or is there something internally that something like `y_pred = torch.sigmoid(model.predict(test_Xs, scaling=(mean,std)))` would not correspond to the class probabilities?

Thanks again for the answer!

Best,
Dominik

**Comment by jan-engelmann** [comment]
> thank you for the answer! We observe that generally performance deteriorates for the mean model when using training statistics. For the trained MixMIL model it actually improves and reduces some of the false positives.

Interesting! Then for the mean model I would remove the reparameterization in your position.

> This I don't understand. For n_trials=1 this should correspond to a logistic regression, right? Or is there something internally that something like y_pred = torch.sigmoid(model.predict(test_Xs, scaling=(mean,std))) would not correspond to the class probabilities?

You're absolutely right. For some reason I assumed you were using 2 trials.

Sounds like a very cool use-case! let me know if I you have any other questions!
